stan/math/prim/meta/require_helpers.hpp

```diff
@@ -0,0 +1,401 @@
+ * Adds unary require aliases that check the `value_type`.
+ * @param check_type The name of the type to check, used to define
+ * `require_<check_type>_t`.
+ * @param checker A struct that returns holds a boolean `value`
+ * @param doxygen_group The doxygen group to add this requires to.
+ */
+#define STAN_ADD_REQUIRE_UNARY_VALUE(check_type, checker, doxygen_group)       \
+  /*! \ingroup doxygen_group */                                                \
+  /*! \addtogroup check_type##_types */                                        \
+  /*! @{ */                                                                    \
+  /*! \brief Require value type satisfies checker */                           \
+  template <typename T>                                                        \
+  using require_##check_type##_vt                                              \
+      = require_t<checker<value_type_t<std::decay_t<T>>>>;                     \
+  /*! \brief Require value type does not satisfy checker */                    \
+  template <typename T>                                                        \
+  using require_not_##check_type##_vt                                          \
+      = require_not_t<checker<value_type_t<std::decay_t<T>>>>;                 \
+  /*! \brief Require all of the value types satisfy checker */                 \
+  template <typename... Types>                                                 \
+  using require_all_##check_type##_vt                                          \
+      = require_all_t<checker<value_type_t<std::decay_t<Types>>>...>;          \
+  /*! \brief Require any of the value types satisfy checker */                 \
+  template <typename... Types>                                                 \
+  using require_any_##check_type##_vt                                          \
+      = require_any_t<checker<value_type_t<std::decay_t<Types>>>...>;          \
+  /*! \brief Require none of the value types satisfy checker */                \
+  template <typename... Types>                                                 \
+  using require_all_not_##check_type##_vt                                      \
+      = require_all_not_t<checker<value_type_t<std::decay_t<Types>>>...>;      \
+  /*! \brief Require at least one of the value types do not satisfy checker */ \
+  template <typename... Types>                                                 \
+  using require_any_not_##check_type##_vt                                      \
+      = require_any_not_t<checker<value_type_t<std::decay_t<Types>>>...>;      \
+/*! @} */
+
+/** \ingroup macro_helpers
+ * Adds binary require aliases.
+ * @param check_type The name of the type to check, used to define
+ * `require_<check_type>_t`.
+ * @param checker A struct that returns holds a boolean `value`
+ * @param doxygen_group The doxygen group to add this requires to.
+ */
+#define STAN_ADD_REQUIRE_BINARY(check_type, checker, doxygen_group)          \
```

Comment:
`STAN_ADD_REQUIRE_BINARY` and value/scalar versions seem to be unused. Is that intentional?

Reply:
They are used in require_generics.hpp for comparison type traits such as 

```cpp
STAN_ADD_REQUIRE_BINARY(same, std::is_same, require_std);
```